### PR TITLE
Lesson 5-7 Part 3: Test Script

### PR DESCRIPTION
**Pull request by Derivability**
Created script to test character device

### CODE REVIEW

**Review by galkinletter — CHANGES_REQUESTED** [review]
02-KernelModules_CharacterDevice/Makefile

Please, add empty line at the end of file

File: 02-KernelModules_CharacterDevice/Makefile [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+else
+
+KERNELDIR := $(BUILD_KERNEL)
+
+all:
+	$(MAKE) -C $(KERNELDIR) M=$(CURDIR) modules
+clean:
+	$(MAKE) -C $(KERNELDIR) M=$(CURDIR) clean
+
+endif
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/Makefile:18]: [02-KernelModules_CharacterDevice/Makefile](https://github.com/Kernel-GL-HRK/gl-kernel-basecamp-2019/pull/106/files#diff-966f94605c34f5523e4fc8ce26c2ca05)

```suggestion
endif

```

Please, add empty line at the end of file

File: 02-KernelModules_CharacterDevice/main.c [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	
+	unregister_chrdev(major, DEVICE_NAME);
+	
+	printk(KERN_INFO "ChD: Module unloaded\n");
+}
+
+module_init(ChD_init);
+module_exit(ChD_exit);
+
+MODULE_LICENSE("GPL");
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/main.c:259]: [02-KernelModules_CharacterDevice/main.c](https://github.com/Kernel-GL-HRK/gl-kernel-basecamp-2019/pull/106/files#diff-389fcb28ab8b9bd0ef8ac68f9a3f772e)

```suggestion
MODULE_LICENSE("GPL");
MODULE_AUTHOR("Name@gmail.com");
MODULE_DESCRIPTION("Modul for test Character Decices");
MODULE_VERSION("1.0")
```
Need fix description of module

Comment (galkinletter) [reply 02-KernelModules_CharacterDevice/main.c:259]: main.c it is not name of driver or module!

Need fix it

**Review by galkinletter — CHANGES_REQUESTED** [review]
Need fix Makefile

Need fix module 

Need fix code  style

File: 02-KernelModules_CharacterDevice/main.c (hunk 2 above)
Comment (Derivability) [reply 02-KernelModules_CharacterDevice/main.c:259]: Please checkout this pull request: https://github.com/Kernel-GL-HRK/gl-kernel-basecamp-2019/pull/105

File: 02-KernelModules_CharacterDevice/main.c [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+		printk(KERN_INFO "ChD: Registered major number %d\n", major);
+	}
+	
+	chdClass = class_create(THIS_MODULE, CLASS_NAME);
+
+	chDevice = device_create(chdClass, NULL, MKDEV(major, 0), NULL, DEVICE_NAME);
+
+	printk(KERN_INFO "ChD: Device initialized succesfully\n");
+
+	buffer_create();
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/main.c:226]: What about handling error from this function?

File: 02-KernelModules_CharacterDevice/main.c [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+	
+	if(major < 0){
+		printk(KERN_ALERT "ChD: Failed to register major number\n");
+		return major;
+	}
+	else{
+		printk(KERN_INFO "ChD: Registered major number %d\n", major);
+	}
+	
+	chdClass = class_create(THIS_MODULE, CLASS_NAME);
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/main.c:220]: Error handling issues below this point.

**Review by oleg-h2o — COMMENTED** [review]
Error handling fixes are required.

File: 02-KernelModules_CharacterDevice/main.c (hunk 4 above)
Comment (Derivability) [reply 02-KernelModules_CharacterDevice/main.c:220]: Please checkout this PR: https://github.com/Kernel-GL-HRK/gl-kernel-basecamp-2019/pull/105

**Pull request closed without merging**